kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ProduceActionRequestSizeLimitIntegrationTest.java

```diff
@@ -0,0 +1,382 @@
+  private static final int TEST_DATA_SIZE = 1024 * 1024;
+  private static final int LIMIT_SIZE = 1024 * 1024 + 8 * 1024;
+
+  private HttpClient httpClient;
+
+  @RegisterExtension
+  public final DefaultKafkaRestTestEnvironment testEnv = new DefaultKafkaRestTestEnvironment(false);
+
+  @BeforeEach
+  public void setUp(TestInfo testInfo) throws Exception {
+    Properties properties = new Properties();
+    // Adding custom KafkaRestConfigs for individual test-cases/test-methods below.
+    if (!testInfo.getDisplayName().contains("ProduceRequestSizeNoLimit")) {
+      properties.put(
+          KafkaRestConfig.PRODUCE_REQUEST_SIZE_LIMIT_MAX_BYTES_CONFIG, String.valueOf(LIMIT_SIZE));
+    }
+    testEnv.kafkaRest().startApp(properties);
+    testEnv.kafkaCluster().createTopic(TEST_TOPIC_NAME, 1, (short) 1);
+
+    SslContextFactory.Client sslContextFactory = new SslContextFactory.Client();
+    sslContextFactory.setSslContext(testEnv.certificates().getSslContext("kafka-rest"));
+    httpClient = new HttpClient(sslContextFactory);
+    httpClient.start();
+  }
+
+  @AfterEach
+  public void tearDown() throws Exception {
+    httpClient.stop();
+    testEnv.kafkaRest().closeApp();
+  }
+
+  @Test
+  @DisplayName("testStreaming_ProduceRequestSizeNoLimit")
+  public void testStreaming_ProduceRequestSizeNoLimit() throws Exception {
+    String clusterId = testEnv.kafkaCluster().getClusterId();
+    URI uri =
+        testEnv
+            .kafkaRest()
+            .target()
+            .path("/v3/clusters/" + clusterId + "/topics/" + TEST_TOPIC_NAME + "/records")
+            .getUri();
+
+    final OutputStreamContentProvider contentProvider = new OutputStreamContentProvider();
+    InputStreamResponseListener responseListener = new InputStreamResponseListener();
+
+    httpClient
+        .POST(uri)
+        .header(HttpHeader.TRANSFER_ENCODING, "chunked")
+        .content(contentProvider, MediaType.APPLICATION_JSON)
+        .send(responseListener); // async request
+
+    httpClient
+        .getExecutor()
+        .execute(
+            () -> {
+              // send data in a separate thread
+              try (OutputStream outputStream = contentProvider.getOutputStream()) {
+                for (int i = 0; i < 5; i++) {
+                  outputStream.write(generateData(TEST_DATA_SIZE).getBytes(StandardCharsets.UTF_8));
+                }
+              } catch (IOException e) {
+                log.error("Error writing to output stream", e);
+              }
+            });
+
+    List<TestProduceResponse> produceResponses = new ArrayList<>();
+    // waiting for response
+    Response response = responseListener.get(1, TimeUnit.MINUTES);
+    if (response.getStatus() == HttpStatus.OK_200) {
+      // Obtain the input stream on the response content
+      try (InputStream input = responseListener.getInputStream()) {
+        // Read the response content as stream
+        BufferedReader reader =
```

Comment:
Is this a duplex way of using streaming in Java?(if yes, exciting)
In parallel send records, and read record-responses on the same http connection :)

Reply:
> Is this a duplex way of using streaming in Java?(if yes, exciting)

yeah, it is the one, where I use Jetty http client (third party library).

There is probably a way to do it with Java Http connection as well but potentially quite complex (something like open output stream and input stream from http connection and use another thread to write to the output stream).